File: go/2022/day11.go
package _2022

import (
	"advent/utils"
	"fmt"
	"regexp"
	"sort"
	"strings"
)

const (
	ADD    = 0
	MULT   = 1
	SQUARE = 2
)

type Operation [2]int
type Monkey struct {
	id        int
	divisor   int
	ifFalse   int
	ifTrue    int
	count     int
	inventory []int
	operation Operation
}
type MonkeyBunch map[int]*Monkey // * means pointer to a monkey

const monkeyRx = `Monkey (\d):
  Starting items: (.*)
  Operation: new = (.*)
  Test: divisible by (\d+)
    If true: throw to monkey (\d)
    If false: throw to monkey (\d)`

func parseOperation(operationStr string) Operation {
	operationStr = strings.TrimSpace(operationStr)
	rxSquare := regexp.MustCompile(`old \* old`)
	rxMult := regexp.MustCompile(`old \* (\d+)`)
	rxAdd := regexp.MustCompile(`old \+ (\d+)`)

	if rxSquare.MatchString(operationStr) {
		return Operation{SQUARE, 69}
	} else if rxMult.MatchString(operationStr) {
		match := rxMult.FindStringSubmatch(operationStr)
		number := utils.ParseInt(match[1])
		return Operation{MULT, number}
	} else if rxAdd.MatchString(operationStr) {
		match := rxAdd.FindStringSubmatch(operationStr)
		number := utils.ParseInt(match[1])
		return Operation{ADD, number}
	} else {
		panic(fmt.Sprintf("Help! Couldn't find any match for: '%v'", operationStr))
	}
}

func parseMonkey(monkeyStr string) Monkey {
	rx := regexp.MustCompile(monkeyRx)
	match := rx.FindStringSubmatch(monkeyStr)

	id := utils.ParseInt(match[1])
	itemsStr := match[2]
	items := []int{}
	for _, item := range strings.Split(itemsStr, ", ") {
		items = append(items, utils.ParseInt(item))
	}
	operationStr := match[3]
	divisor := utils.ParseInt(match[4])
	ifTrue := utils.ParseInt(match[5])
	ifFalse := utils.ParseInt(match[6])

	operation := parseOperation(operationStr)
	return Monkey{
		id:        id,
		operation: operation,
		ifTrue:    ifTrue,
		ifFalse:   ifFalse,
		divisor:   divisor,
		inventory: items,
	}
}

func ParseMonkeys(input string) MonkeyBunch {
	monkeyBunch := MonkeyBunch{}
	for id, monkeyStr := range strings.Split(input, "\n\n") {
		monkey := parseMonkey(monkeyStr)
		monkeyBunch[id] = &monkey
	}
	return monkeyBunch
}

func monkeyThrow(monkeyId int, item int, others MonkeyBunch, decreaseWorry func(int) int) {
	monkey := others[monkeyId]
	number := monkey.operation[1]
	switch monkey.operation[0] {
	case SQUARE:
		item *= item
	case MULT:
		item *= number
	case ADD:
		item += number
	}
	item = decreaseWorry(item)
	otherId := monkey.ifFalse
	if item%monkey.divisor == 0 {
		otherId = monkey.ifTrue
	}
	others[otherId].inventory = append(others[otherId].inventory, item)
	monkey.count++
}

func getMostActiveMonkeys(monkeys MonkeyBunch, top int) []Monkey {
	sortedMonkeys := []Monkey{}
	for _, monkeyPtr := range monkeys {
		sortedMonkeys = append(sortedMonkeys, *monkeyPtr)
	}
	sort.Slice(
		sortedMonkeys,
		func(i, j int) bool {
			return sortedMonkeys[i].count > sortedMonkeys[j].count
		},
	)
	return sortedMonkeys[:top]
}

func Day11Part1(input string) string {
	monkeys := ParseMonkeys(input)
	worryFunc := func(number int) int { return number / 3 }
	for round := 0; round < 20; round++ {
		for monkeyId := 0; monkeyId < len(monkeys); monkeyId++ {
			monkey := monkeys[monkeyId]
			for _, item := range monkey.inventory {
				monkeyThrow(monkeyId, item, monkeys, worryFunc)
			}
			monkey.inventory = []int{}
		}
	}
	top2 := getMostActiveMonkeys(monkeys, 2)
	result := top2[0].count * top2[1].count
	return fmt.Sprint(result)
}

func Day11Part2(input string) string {
	monkeys := ParseMonkeys(input)
	modulus := 1
	for _, monkey := range monkeys {
		modulus *= monkey.divisor
	}
	worryFunc := func(number int) int { return number % modulus }
	for round := 0; round < 10000; round++ {
		for monkeyId := 0; monkeyId < len(monkeys); monkeyId++ {
			monkey := monkeys[monkeyId]
			for _, item := range monkey.inventory {
				monkeyThrow(monkeyId, item, monkeys, worryFunc)
			}
			monkey.inventory = []int{}
		}
	}
	top2 := getMostActiveMonkeys(monkeys, 2)
	result := top2[0].count * top2[1].count
	return fmt.Sprint(result)
}

File: python/puzzles/y2022/day11.py
import re
from dataclasses import dataclass
from enum import Enum
from functools import reduce
from typing import Callable

import utils

example = """Monkey 0:
  Starting items: 79, 98
  Operation: new = old * 19
  Test: divisible by 23
    If true: throw to monkey 2
    If false: throw to monkey 3

Monkey 1:
  Starting items: 54, 65, 75, 74
  Operation: new = old + 6
  Test: divisible by 19
    If true: throw to monkey 2
    If false: throw to monkey 0

Monkey 2:
  Starting items: 79, 60, 97
  Operation: new = old * old
  Test: divisible by 13
    If true: throw to monkey 1
    If false: throw to monkey 3

Monkey 3:
  Starting items: 74
  Operation: new = old + 3
  Test: divisible by 17
    If true: throw to monkey 0
    If false: throw to monkey 1"""


monkey_rx = """Monkey (\d):
  Starting items: (.*)
  Operation: new = (.*)
  Test: divisible by (\d+)
    If true: throw to monkey (\d)
    If false: throw to monkey (\d)"""


class Operation(Enum):
    ADD = 0
    MULT = 1
    SQUARE = 2


@dataclass
class Monkey:
    inventory: list[int]
    operation: tuple[Operation, int]
    divisor: int
    if_false: int
    if_true: int
    count: int = 0


MonkeyBunch = dict[int, Monkey]


def parse_monkey(monkey_str: str) -> Monkey:
    rx = re.compile(monkey_rx)
    match = rx.search(monkey_str)
    id, items, operation, divisor, if_true, if_false = match.groups()
    match operation.split():
        case ["old", "*", "old"]:
            operation = (Operation.SQUARE, 69)
        case ["old", "*", number]:
            operation = (Operation.MULT, int(number))
        case ["old", "+", number]:
            operation = (Operation.ADD, int(number))

    return Monkey(
        inventory=list(map(int, items.split(", "))),
        operation=operation,
        divisor=int(divisor),
        if_true=int(if_true),
        if_false=int(if_false),
    )


def parse_input(raw: str) -> MonkeyBunch:
    return {ii: parse_monkey(monkey_str) for ii, monkey_str in enumerate(raw.split("\n\n"))}


def throw(monkey: Monkey, item: int, monkey_bunch: MonkeyBunch, decrease_worry: Callable):
    match monkey.operation:
        case [Operation.SQUARE, _]:
            item *= item
        case [Operation.MULT, number]:
            item *= number
        case [Operation.ADD, number]:
            item += number
    item = decrease_worry(item)
    result = item % monkey.divisor == 0
    other_id = monkey.if_true if result else monkey.if_false
    other = monkey_bunch[other_id]
    other.inventory.append(item)
    monkey.count += 1


@utils.profile
def part1(raw: str):
    monkey_bunch = parse_input(raw)
    for round in range(20):
        for monkey in monkey_bunch.values():
            for item in monkey.inventory:
                throw(
                    monkey=monkey,
                    item=item,
                    monkey_bunch=monkey_bunch,
                    decrease_worry=lambda number: number // 3,
                )
            monkey.inventory = []

    most_active = sorted(monkey_bunch.values(), key=lambda x: -x.count)
    return most_active[0].count * most_active[1].count


@utils.profile
def part2(raw: str):
    """
    Use clock arithmetic to keep the worry numbers low, while still allowing the "is divisible by
    x" calculations to give the same result. For all monkeys' calculations to be unaffected,
    the max value of the clock should be the lowest common multiple of all the monkeys' divisors.
    Because the divisors are all prime, the lowest common multiple is simply the product.
    """

    monkey_bunch = parse_input(raw)
    modulus = reduce(lambda a, b: a * b, (m.divisor for m in monkey_bunch.values()))
    for round in range(10000):
        for monkey in monkey_bunch.values():
            for item in monkey.inventory:
                throw(
                    monkey=monkey,
                    item=item,
                    monkey_bunch=monkey_bunch,
                    decrease_worry=lambda number: number % modulus,
                )
            monkey.inventory = []

    most_active = sorted(monkey_bunch.values(), key=lambda x: -x.count)
    return most_active[0].count * most_active[1].count


if __name__ == "__main__":
    raw = utils.load_puzzle_input("2022/day11")
    assert part1(raw) == 64032
    assert part2(raw) == 12729522272
